src/Elasticsearch/Extension/SearchExtension.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the RollerworksSearch package.
+ *
+ * (c) Sebastiaan Stok <s.stok@rollerscapes.net>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Rollerworks\Component\Search\ApiPlatform\Elasticsearch\Extension;
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\QueryCollectionExtensionInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Elastica\Client;
+use Elastica\Document;
+use Elastica\Search;
+use Rollerworks\Component\Search\ApiPlatform\ArrayKeysValidator;
+use Rollerworks\Component\Search\Elasticsearch\ElasticsearchFactory;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Class SearchExtension.
+ */
+class SearchExtension implements QueryCollectionExtensionInterface
+{
+    private $requestStack;
+    private $elasticsearchFactory;
+    private $elasticaClient;
+
+    public function __construct(RequestStack $requestStack, ElasticsearchFactory $elasticsearchFactory, Client $elasticaClient)
+    {
+        $this->requestStack = $requestStack;
+        $this->elasticsearchFactory = $elasticsearchFactory;
+        $this->elasticaClient = $elasticaClient;
+    }
+
+    public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null)
+    {
+        $request = $this->requestStack->getCurrentRequest();
+
+        /** @var SearchCondition $condition */
+        if (!$request || null === $condition = $request->attributes->get('_api_search_condition')) {
+            return;
+        }
+
+        $context = $request->attributes->get('_api_search_context');
+        $configuration = $request->attributes->get('_api_search_config');
+        $configPath = "{$resourceClass}#attributes[rollerworks_search][contexts][{$context}][elasticsearch]";
+
+        if (empty($configuration['elasticsearch'])) {
+            return;
+        }
+
+        $configuration = (array) $configuration['elasticsearch'];
+        ArrayKeysValidator::assertOnlyKeys($configuration, ['mappings'], $configPath);
+
+        $ids =
+
+        // this snippet looks weird, factory should create a
+        $conditionGenerator = $this->elasticsearchFactory->createCachedConditionGenerator(
+            $this->elasticsearchFactory->createConditionGenerator($condition)
+        );
+
+        foreach ($configuration['mappings'] as $fieldName => $mapping) {
+            $conditionGenerator->registerField($fieldName, $mapping);
+        }
+
+        // TODO: temporary, how to do this better?
+        $identifier = 'id';
+        $search = new Search($this->elasticaClient);
```

Comment:
@sstok this is where the question of responsibility comes in, if you have any suggestions, I'd love to hear them.

Reply:
This should probably be taken from entity metadata. Don't know about composite keys.

Reply:
Problem with this approach is you embed the search functionality into a API Platform extension.

What would be better is to have it be available stand-alone, probably with what FOS Elastica is doing, repository based? Ie. repository implementing an interface with

```php
public function search(SearchCondition $searchCondition): array;
```

with a trait converting the condition using Elasticsearch query generator and returning proper results (as Entities).

Reply:
Hmm. I agree this functionality is better suited in a separate class/package, but it's properly easiest (for now) to keep it in this package and see what works and can be abstracted.

Reply:
@sstok So what do you think about this as it is? What changes would you like to see to get this merged?